code/_onclick/click.dm

```diff
@@ -1,145 +1,231 @@
+	if(flags_atom & INITIALIZED)
+		SEND_SIGNAL(src, COMSIG_CLICK, location, control, params, usr)
+		usr.ClickOn(src, params)
+
+
+/atom/DblClick(location, control, params)
+	if(flags_atom & INITIALIZED)
+		usr.DblClickOn(src, params)
+
 
-/client/Click(atom/A, location, control, params)
-	if(control)	// No .click macros allowed
-		return usr.do_click(A, location, params)
+/atom/MouseWheel(delta_x, delta_y, location, control, params)
+	if(flags_atom & INITIALIZED)
+		usr.MouseWheelOn(src, delta_x, delta_y, params)
 
 
-/mob/proc/do_click(atom/A, location, params)
+/*
+	Standard mob ClickOn()
+	Handles exceptions: Buildmode, middle click, modified clicks, mech actions
+
+	After that, mostly just check your state, check whether you're holding an item,
+	check whether you're adjacent to the target, then pass off the click to whoever
+	is receiving it.
+	The most common are:
+	* mob/UnarmedAttack(atom,adjacent) - used here only when adjacent, with no item in hand; in the case of humans, checks gloves
+	* atom/attackby(item,user) - used only when adjacent
+	* item/afterattack(atom,user,adjacent,params) - used both ranged and adjacent
+	* mob/RangedAttack(atom,params) - used only ranged, only used for tk and laser eyes but could be changed
+*/
+/mob/proc/ClickOn(atom/A, params)
 	if(world.time <= next_click)
 		return
 	next_click = world.time + 1
 
 	if(check_click_intercept(params, A))
 		return
 
-	// No clicking on atoms with the NOINTERACT flag
-	if((A.flags_atom & NOINTERACT))
-		if(istype(A, /obj/screen/click_catcher))
-			var/list/mods = params2list(params)
-			var/turf/TU = params2turf(mods["screen-loc"], get_turf(usr.client ? usr.client.eye : usr), usr.client)
-			if(TU)
-				do_click(TU, location, params)
+	if(notransform)
 		return
 
-
-	var/list/mods = params2list(params)
-
-	// Don't allow any other clicks while dragging something
-	if(mods["drag"])
+	if(SEND_SIGNAL(src, COMSIG_MOB_CLICKON, A, params) & COMSIG_MOB_CANCEL_CLICKON)
 		return
 
-	var/click_handled = 0
-	// Click handled elsewhere. (These clicks are not affected by the next_move cooldown)
-	click_handled = click(A, mods)
-	click_handled |= A.clicked(src, mods, location, params)
-
-	if(click_handled)
+	var/list/modifiers = params2list(params)
+	if(modifiers["shift"] && modifiers["middle"])
+		ShiftMiddleClickOn(A)
+		return
+	if(modifiers["shift"] && modifiers["ctrl"])
+		CtrlShiftClickOn(A)
+		return
+	if(modifiers["ctrl"] && modifiers["middle"])
+		CtrlMiddleClickOn(A)
+		return
+	if(modifiers["middle"])
+		MiddleClickOn(A)
+		return
+	if(modifiers["shift"])
+		ShiftClickOn(A)
+		return
+	if(modifiers["alt"])
+		AltClickOn(A)
+		return
+	if(modifiers["ctrl"])
+		CtrlClickOn(A)
 		return
 
-	// Default click functions from here on.
 	if(is_mob_incapacitated(TRUE))
 		return
 
 	face_atom(A)
 
-	// Special type of click.
-	if(is_mob_restrained())
-		RestrainedClickOn(A)
+	if(next_move > world.time)
+		return
+
+	if(istype(loc, /obj/vehicle/multitile/root/cm_armored))
```

Comment:
Define `click_action()` at the atom level, to return FALSE or null, modify the other definitions to always return TRUE (or invert that logic), and do an `if(loc.click_action(A, src, params)) then return`

Reply:
Again, /tg/ code except mech is replaced with a tank :c
But yeah, sure